Dockerfile

```diff
@@ -1,31 +1,37 @@
-FROM python:3.8-slim-buster as builder
+FROM python:3.9-slim as base
 
 ENV LANG=C.UTF-8 \
     DEBIAN_FRONTEND=noninteractive \
     PIP_NO_CACHE_DIR=true
 
-RUN apt-get update && apt-get install --no-install-recommends -qq -y gcc g++ && apt-get clean autoclean \
-  && apt-get autoremove -y \
-  && rm -rf /var/lib/apt/lists/* \
-  && rm -f /var/cache/apt/archives/*.deb
+RUN apt-get update && apt-get install -y --no-install-recommends -qq gcc=4:10.2.1-1 g++=4:10.2.1-1 curl=7.74.0-1.3+deb11u1 \
+ && apt-get clean && rm -rf /var/lib/apt/lists/*
 
-COPY requirements.txt .
-RUN pip install --upgrade pip && pip install --trusted-host pypi.python.org -r requirements.txt
+WORKDIR /app
 
-FROM python:3.8-slim-buster as production
+FROM base as builder
 
-ENV PYTHONDONTWRITEBYTECODE 1
-ENV PYTHONUNBUFFERED 1
+ENV POETRY_VERSION=1.1.13
+RUN pip install --no-cache-dir poetry==$POETRY_VERSION
 
-WORKDIR /app
+COPY pyproject.toml poetry.lock ./
+RUN python -m venv --copies /venv
+
+RUN . /venv/bin/activate && poetry install --no-dev --no-root
+
+FROM base as production
+
+COPY --from=builder /venv /venv
 
 RUN mkdir -p /var/www && chown www-data /var/www && \
-    apt-get update && apt-get install --no-install-recommends -qq -y curl && \
-    apt-get clean && find /var/lib/apt/lists/ -type f -delete && \
-    chown www-data /app/
+    chown -R www-data /app/ && chown -R www-data /venv
 
-ENV PATH=$PATH:/usr/local/bin
-ENV PYTHONPATH="/usr/local/lib/python3.8/site-packages/"
+ENV PYTHONPATH="/venv/lib/python3.9/site-packages/"
```

Comment:
It's common question. Why do we need to hard code docker envs? In most case we prepared .env config which will be uploaded into bash profile on deployed machine.

So, I wanted to highlight that if we posted them in .env file => ci/di could provide them into docker setup/config and etc